File: frontend/src/lib/board/geometry.ts
// TS mirror of the Rust `BoardGeometry` published by `Rules::board_geometry()`
// and exposed via WASM `get_board_geometry(rules_name)`. These types must stay
// in sync with `sorry-core/src/geometry.rs`.
//
// Serialization notes (pinned by
// `sorry-core/src/rules.rs::board_geometry_json_shape_is_frontend_friendly`):
//   - `SpaceId(u32)` and `PlayerId(u8)` are newtype structs — they serialize
//     transparently as bare numbers.
//   - `Space` enum uses default externally-tagged form: `{"Track": 0}`,
//     `{"StartArea": 2}`, `{"Safety": [0, 3]}`, `{"Home": 1}`.

export type SpaceId = number;
export type PlayerId = number;

export type SpaceKind =
	| { Track: number }
	| { StartArea: PlayerId }
	| { Safety: [PlayerId, number] }
	| { Home: PlayerId };

export interface SpaceLayout {
	id: SpaceId;
	kind: SpaceKind;
	/** `[x, y]` in the board-local `[-1, 1]` square. */
	center: [number, number];
	/** Pawn facing direction, counter-clockwise from `+x` in degrees. */
	tangent_deg: number;
	/** Indexed by `PlayerId`. `null` means no neighbor for that player. */
	forward: (SpaceId | null)[];
	/** Indexed by `PlayerId`. `null` means no neighbor for that player. */
	backward: (SpaceId | null)[];
}

export interface SlideLayout {
	head: SpaceId;
	end: SpaceId;
	path: SpaceId[];
	owner: PlayerId;
}

export interface PlayerLayout {
	player: PlayerId;
	start_area: SpaceId;
	start_exit: SpaceId;
	home: SpaceId;
}

export interface BoardGeometry {
	/** `[xmin, ymin, xmax, ymax]` — normalized board-local bounds. */
	bounds: [number, number, number, number];
	spaces: SpaceLayout[];
	slides: SlideLayout[];
	/** Indexed by `PlayerId`. */
	players: PlayerLayout[];
}

// ─── Helpers ──────────────────────────────────────────────────────────

export function spaceById(geom: BoardGeometry, id: SpaceId): SpaceLayout {
	for (const s of geom.spaces) {
		if (s.id === id) return s;
	}
	throw new Error(`unknown SpaceId: ${id}`);
}

/**
 * Convert a normalized `[-1, 1]` center to planar world-space `(x, z)`.
 * `y` (vertical) is left for the renderer to add. `scale` multiplies both
 * planar axes — default `2` puts the board in a `4×4` world square.
 */
export function toPlanar(center: [number, number], scale = 2.0): [number, number] {
	return [center[0] * scale, center[1] * scale];
}

export function isTrack(kind: SpaceKind): kind is { Track: number } {
	return 'Track' in kind;
}

export function isStartOf(kind: SpaceKind, player: PlayerId): boolean {
	return 'StartArea' in kind && kind.StartArea === player;
}

export function isSafetyOf(kind: SpaceKind, player: PlayerId): boolean {
	return 'Safety' in kind && kind.Safety[0] === player;
}

export function isHomeOf(kind: SpaceKind, player: PlayerId): boolean {
	return 'Home' in kind && kind.Home === player;
}

File: sorry-core/src/geometry.rs
//! Board geometry published by `Rules`. Single source of truth for where each
//! `SpaceId` lives in normalized `[-1, 1]` board-local coordinates, plus the
//! per-player adjacency graph a client walks for hop-by-hop animation without
//! crossing the WASM boundary.

use serde::{Deserialize, Serialize};

use crate::board::{PlayerId, Space, SpaceId};

/// Layout for a single space. `center` is in the board-local `[-1, 1]` square.
/// `tangent_deg` is the direction a pawn on this space faces, measured
/// counter-clockwise from `+x` in degrees.
///
/// `forward[i]` / `backward[i]` is the per-player adjacency for `PlayerId(i)`.
/// `None` means the neighbor is off the board for that player (e.g. `Home`
/// has no forward neighbor; another player's `Safety` is unreachable).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SpaceLayout {
    pub id: SpaceId,
    pub kind: Space,
    pub center: [f32; 2],
    pub tangent_deg: f32,
    pub forward: Vec<Option<SpaceId>>,
    pub backward: Vec<Option<SpaceId>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SlideLayout {
    pub head: SpaceId,
    pub end: SpaceId,
    pub path: Vec<SpaceId>,
    pub owner: PlayerId,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PlayerLayout {
    pub player: PlayerId,
    pub start_area: SpaceId,
    pub start_exit: SpaceId,
    pub home: SpaceId,
}

/// Full board layout snapshot — static per `Rules` impl, immutable for the
/// life of a game. `bounds` is `[xmin, ymin, xmax, ymax]` of the usable
/// layout region.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BoardGeometry {
    pub bounds: [f32; 4],
    pub spaces: Vec<SpaceLayout>,
    pub slides: Vec<SlideLayout>,
    pub players: Vec<PlayerLayout>,
}
